File: bot/bot.py
import os
import sys
import logging
import asyncio
from telegram.ext import Application

# Add parent directory to path for imports
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from config import TELEGRAM_BOT_TOKEN, BOT_VERSION
from database import init_db
from .handlers import setup_handlers
from trading.copytrade import start_monitoring as start_copytrade_monitoring
from trading.alerts import start_alert_monitoring

# Setup logging
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    level=logging.INFO
)
logger = logging.getLogger(__name__)

async def main():
    """Start the bot."""
    # Initialize database
    init_db()
    
    # Create the Application
    application = Application.builder().token(TELEGRAM_BOT_TOKEN).build()
    
    # Setup handlers
    setup_handlers(application)
    
    # Start the Bot
    logger.info(f"Starting Zoracle Bot v{BOT_VERSION}...")
    await application.initialize()
    await application.start_polling()
    
    # Start monitoring services
    asyncio.create_task(start_copytrade_monitoring())
    asyncio.create_task(start_alert_monitoring(application.bot))
    
    logger.info("Bot is running. Press Ctrl+C to stop.")
    
    # Run the bot until the user presses Ctrl-C
    try:
        await application.updater.start_polling()
    finally:
        await application.updater.stop()
        await application.stop()

def run_bot():
    """Run the bot."""
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        logger.info("Bot stopped by user.")
    except Exception as e:
        logger.error(f"Error running bot: {str(e)}")

if __name__ == '__main__':
    run_bot() 

File: bot/bot.js
const TelegramBot = require('node-telegram-bot-api');
const { CONFIG } = require('../config');
const { users, token } = require('./baseBot');

// Initialize bot with polling
let bot;
try {
  bot = new TelegramBot(token, { 
    polling: true,
    filepath: false // Disable file downloading to avoid disk space issues
  });
  
  // Add isPolling method to check if bot is polling
  bot.isPolling = function() {
    return this.polling;
  };
  
  // Add specific error handlers
  bot.on('polling_error', (error) => {
    if (error.code === 'ETELEGRAM' && error.message.includes('409 Conflict')) {
      console.error('❌ Multiple bot instances detected!');
      console.error('Make sure only one instance of the bot is running.');
      console.error('Try running: pkill -f "node index.js" before starting again.');
      process.exit(1);
    } else if (error.code === 'ETELEGRAM' && error.message.includes('401 Unauthorized')) {
      console.error('❌ Invalid Telegram bot token!');
      console.error('Please check your TELEGRAM_BOT_TOKEN in the .env file.');
      console.error('You can get a valid token from @BotFather on Telegram.');
      process.exit(1);
    } else {
      console.error('Polling error:', error.code, error.message);
    }
  });
} catch (error) {
  console.error('❌ Failed to initialize Telegram bot:', error.message);
  process.exit(1);
}

// Logging Middleware
bot.on('message', (msg) => {
  console.log(`[${new Date().toISOString()}] Chat ID: ${msg.chat.id}, User: ${msg.from.username || msg.from.first_name}, Message: ${msg.text}`);
});

// Start Command
bot.onText(/\/start/, (msg) => {
  const chatId = msg.chat.id;
  bot.sendMessage(chatId, `Welcome! I am your Zoracle Telegram Bot. How can I assist you today?`);
});

// Help Command
bot.onText(/\/help/, (msg) => {
  const chatId = msg.chat.id;
  const helpMessage = `
Here are the commands you can use:
/start - Start the bot and get a welcome message.
/help - Display this help message.
/wallet - Manage your wallet (create, import, view balance).
/balance - Check your ETH and token balances.
/buy <token_address> <amount> - Buy tokens with ETH.
/sell <token_address> <amount> - Sell tokens for ETH.
/portfolio - View your token portfolio.
/transactions - View your transaction history.
/pnl - Calculate profit/loss for your holdings.
/search <query> - Search for tokens.
/trending - Show trending tokens.
/new - Show newly created tokens.
/alerts - Manage your price alerts.
/addalert <token_address> - Add a price alert.
/mirror <wallet_address> - Mirror another wallet's trades.
/mirrors - List active mirrors.
`;
  bot.sendMessage(chatId, helpMessage);
});

// About Command
bot.onText(/\/about/, (msg) => {
  const chatId = msg.chat.id;
  const aboutMessage = `
Zoracle Telegram Bot v${CONFIG.BOT_VERSION}

A Telegram bot for trading Zora content tokens on Base blockchain with advanced features including wallet management, swap engine, portfolio tracking, content discovery, alerts, and copy-trading.

Built with ❤️ for the Base ecosystem.
`;
  bot.sendMessage(chatId, aboutMessage);
});

// Import command handlers
require('./handlers/walletHandlers')(bot, users);
require('./handlers/tradeHandlers')(bot, users);
require('./handlers/portfolioHandlers')(bot, users);
require('./handlers/discoveryHandlers')(bot, users);
require('./handlers/alertHandlers')(bot, users);
require('./handlers/copytradeHandlers')(bot, users);

module.exports = { bot, users }; 